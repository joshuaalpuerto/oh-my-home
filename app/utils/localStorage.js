import localForage from 'localforage'

/**
 * Please wrap this instance to a try/catch block
 */

export async function setItem (...args) {
  return await localForage.setItem(...args) // eslint-disable-line
}

export async function getItem (key) {
  return await localForage.getItem(key) // eslint-disable-line
}

export async function removeItem (key) {
  return await localForage.removeItem(key) // eslint-disable-line
}
