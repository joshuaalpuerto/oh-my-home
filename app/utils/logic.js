/**
 * Currying for instead of using *ugly SWITCH statement
 * @param {*} cases
 */
export const switchFn = cases => defaultCase => key =>
key in cases ? cases[key] : defaultCase
