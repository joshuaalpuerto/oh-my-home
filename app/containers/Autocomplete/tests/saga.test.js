
import { put, call } from 'redux-saga/effects'
import { cloneableGenerator } from 'redux-saga/utils'
import { setItem } from 'utils/localStorage'
import SearchFixture from 'fixtures/search.json'
import {
  RECENT_SEARCH
} from 'containers/App/constants'

import autocompleteSagas, {
  fetchAutocompleteSagas,
  fetchAutocomplete,
  fetchRecentSearchedSaga,
  fetchRecentSearched,
  updateRecentSearchedSaga,
  updateRecentSearched
} from '../saga'
import {
  setAutoCompleteAction,
  setRecentSearchAction
} from '../actions'

const args = {}
const results = SearchFixture

describe('Autocomplete Saga', () => {
  describe('fetchAutocompleteSagas', () => {
    let fetchAutocompleteSagasMock

    beforeEach(() => {
      fetchAutocompleteSagasMock = fetchAutocompleteSagas()
      const callFetchLatest = fetchAutocompleteSagasMock.next().value
      expect(callFetchLatest).toMatchSnapshot()
    })

    describe('fetchAutocomplete', () => {
      // pass empty object for params.
      const fetchAutocompleteMock = fetchAutocomplete(args)

      it('should match the snapshot for setting token', () => {
        const callingAPI = fetchAutocompleteMock.next().value
        expect(callingAPI).toMatchSnapshot()
      })

      it('should set setAutoCompleteAction', () => {
        // eslint-disable-next-line no-unused-vars
        const settingData = fetchAutocompleteMock.next(results).value
        expect(settingData).toEqual(
          put(setAutoCompleteAction(results.predictions))
        )
      })
    })
  })

  describe('fetchRecentSearchedSaga', () => {
    let fetchRecentSearchedSagaMock

    beforeEach(() => {
      fetchRecentSearchedSagaMock = fetchRecentSearchedSaga()
      const callFetchLatest = fetchRecentSearchedSagaMock.next().value
      expect(callFetchLatest).toMatchSnapshot()
    })

    describe('fetchRecentSearched', () => {
      // pass empty object for params.
      const fetchRecentSearchedMock = cloneableGenerator(fetchRecentSearched)()

      it('should match the snapshot getting from storage', () => {
        const callingAPI = fetchRecentSearchedMock.next().value
        expect(callingAPI).toMatchSnapshot()
      })

      it('should set setRecentSearchAction with result', () => {
        // eslint-disable-next-line no-unused-vars
        const mock = fetchRecentSearchedMock.clone()
        const settingData = mock.next(results).value
        expect(settingData).toEqual(
          put(setRecentSearchAction(results))
        )
      })

      it('should set setRecentSearchAction with empty', () => {
        // eslint-disable-next-line no-unused-vars
        const mock = fetchRecentSearchedMock.clone()
        const settingData = mock.next().value
        expect(settingData).toEqual(
          put(setRecentSearchAction([]))
        )
      })
    })
  })

  describe('updateRecentSearchedSaga', () => {
    let updateRecentSearchedSagaMock

    beforeEach(() => {
      updateRecentSearchedSagaMock = updateRecentSearchedSaga()
      const callFetchLatest = updateRecentSearchedSagaMock.next().value
      expect(callFetchLatest).toMatchSnapshot()
    })

    describe('updateRecentSearched', () => {
      const payload = {
        value: 1,
        description: 'singapore'
      }
      const storage = [{
        value: 2,
        description: 'singapore, singapore'
      }]
      // pass empty object for params.
      const updateRecentSearchedMock = cloneableGenerator(updateRecentSearched)({ payload })

      it('should match the snapshot getting from storage', () => {
        const callingAPI = updateRecentSearchedMock.next().value
        expect(callingAPI).toMatchSnapshot()
      })

      it('should match the setting to storage', () => {
        const settingStorageMock = updateRecentSearchedMock.clone()

        const callingAPI = settingStorageMock.next(storage).value
        expect(callingAPI).toEqual(
          call(setItem, RECENT_SEARCH, [payload, ...storage])
        )
      })

      it('should match the only to payload', () => {
        const settingStorageMock = updateRecentSearchedMock.clone()

        const callingAPI = settingStorageMock.next(null).value
        expect(callingAPI).toEqual(
          call(setItem, RECENT_SEARCH, [payload])
        )
      })

      it('should call fetchRecentSearched()', () => {
        // eslint-disable-next-line no-unused-vars
        const settingStorage = updateRecentSearchedMock.next().value
        const callingAPI = updateRecentSearchedMock.next().value
        expect(callingAPI).toMatchSnapshot()
      })
    })
  })

  describe('defaultGenerators', () => {
    const generator = autocompleteSagas()

    it('should have 3 functions', () => {
      const watchers = generator.next().value
      expect(watchers.length).toEqual(3)
    })
  })
})
