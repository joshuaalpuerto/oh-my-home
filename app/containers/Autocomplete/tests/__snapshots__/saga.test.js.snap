// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Saga fetchAutocompleteSagas fetchAutocomplete should match the snapshot for setting token 1`] = `
Object {
  "@@redux-saga/HELPER": true,
  "name": "takeLatest(app/Autocomplete/GET_AUTOCOMPLETE, fetchAutocomplete)",
  "next": [Function],
  "return": [Function],
  "throw": [Function],
  Symbol(Symbol.iterator): [Function],
}
`;

exports[`Autocomplete Saga fetchAutocompleteSagas fetchAutocomplete should match the snapshot for setting token 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://qa.omh.sg/test/autocomplete?input=undefined",
      Object {
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`Autocomplete Saga fetchAutocompleteSagas fetchAutocomplete should set setAutoCompleteAction 1`] = `
Object {
  "@@redux-saga/HELPER": true,
  "name": "takeLatest(app/Autocomplete/GET_AUTOCOMPLETE, fetchAutocomplete)",
  "next": [Function],
  "return": [Function],
  "throw": [Function],
  Symbol(Symbol.iterator): [Function],
}
`;
